### PR TITLE
feat: address collection for one click widgets

### PR DESCRIPTION
**Pull request by ArushKapoorJuspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Added Support to collect billing and shipping details from ApplePay, GooglePay and Paypal instead of the SDK

## How did you test it?

https://docs.google.com/document/d/1yvMroZvHGQQrPwMg5WvjEd7kxRFWslpJzzdFsSQTXUk/edit?usp=sharing

## Checklist

- [x] I ran `npm run re:build`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: src/Payments/ApplePay.res [hunk 1]
```diff
@@ -302,13 +285,41 @@ let make = (
         if dict->Dict.get("applePayProcessPayment")->Option.isSome {
           let token =
             dict->Dict.get("applePayProcessPayment")->Option.getOr(Dict.make()->JSON.Encode.object)
+
+          let billingContact =
+            dict
+            ->Dict.get("applePayBillingContact")
```
Comment (PritishBudhiraja) [inline src/Payments/ApplePay.res:291]: use `getDictfromDict`

File: src/Payments/ApplePay.res [hunk 2]
```diff
@@ -302,13 +285,41 @@ let make = (
...
+            ->Option.flatMap(JSON.Decode.object)
+            ->Option.getOr(Dict.make())
+            ->ApplePayTypes.billingContactItemToObjMapper
+
+          let shippingContact =
+            dict
+            ->Dict.get("applePayShippingContact")
```
Comment (PritishBudhiraja) [inline src/Payments/ApplePay.res:298]: use `getDictfromDict`

File: src/Payments/ApplePay.res [hunk 3]
```diff
@@ -302,13 +285,41 @@ let make = (
...
+            ~statesList=stateJson,
+          )
+
           let bodyDict = PaymentBody.applePayBody(~token, ~connectors)
-          processPayment(bodyDict, ())
+
+          let applePayBody =
+            bodyDict
+            ->Dict.fromArray
+            ->JSON.Encode.object
```
Comment (PritishBudhiraja) [inline src/Payments/ApplePay.res:315]: `getJsonFromArrayOfJson`

File: src/Payments/GPay.res [hunk 4]
```diff
@@ -79,12 +70,25 @@ let make = (
...
 
+  let (stateJson, setStatesJson) = React.useState(_ => JSON.Encode.null)
+
+  React.useEffect0(() => {
+    AddressPaymentInput.importStates("./../States.json")
+    ->then(res => {
+      setStatesJson(_ => res.states)
+      resolve()
+    })
+    ->ignore
```
Comment (PritishBudhiraja) [inline src/Payments/GPay.res:87]:
```
    AddressPaymentInput.importStates("./../States.json")
    ->then(res => {
      setStatesJson(_ => res.states)
      resolve()
    })
```

This code is repetitive.

File: src/Payments/GPay.res [hunk 5]
```diff
@@ -102,6 +106,33 @@ let make = (
           ~body=PaymentBody.gpayBody(~payObj=obj, ~connectors),
         )
 
+        let billingContact =
+          obj.paymentMethodData.info
+          ->getDictFromJson
+          ->Utils.getJsonObjectFromDict("billingAddress")
```
Comment (PritishBudhiraja) [inline src/Payments/GPay.res:112]: Utils. can be removed already open

File: src/Payments/GPay.res [hunk 6]
```diff
@@ -102,6 +106,33 @@ let make = (
...
+          ->getDictFromJson
+          ->billingContactItemToObjMapper
+
+        let shippingContact =
+          metadata
+          ->getDictFromJson
+          ->Utils.getJsonObjectFromDict("shippingAddress")
```
Comment (PritishBudhiraja) [inline src/Payments/GPay.res:119]: Utils. can be removed already open

File: src/Payments/GPay.res [hunk 7]
```diff
@@ -102,6 +106,33 @@ let make = (
...
+          ->getDictFromJson
+          ->billingContactItemToObjMapper
+
+        let email =
+          metadata
+          ->getDictFromJson
+          ->Utils.getString("email", "")
```
Comment (PritishBudhiraja) [inline src/Payments/GPay.res:126]: Utils. can be removed already open

File: src/Payments/PaymentMethodsRecord.res [hunk 8]
```diff
@@ -526,21 +533,31 @@ let getPaymentMethodsFieldTypeFromString = (str, isBancontact) => {
...
-    )
+  | "ALL" => {
+      let countryArr = Country.country->Array.map(item => item.countryName)
+      isAddressCountry ? AddressCountry(countryArr) : ShippingAddressCountry(countryArr)
+    }
+  | _ => {
+      let countryArr =
+        Country.country
+        ->Array.filter(item => options->Array.includes(item.isoAlpha2))
+        ->Array.map(item => item.countryName)
```
Comment (PritishBudhiraja) [inline src/Payments/PaymentMethodsRecord.res:558]: Can be reduce to a single loop- 

```
const filteredCountries = Country.country.reduce((accumulator, currentItem) => {
    if (options.includes(currentItem.isoAlpha2)) {
        accumulator.push(currentItem.countryName);
    }
    return accumulator;
}, []);
```

File: src/Utilities/Utils.res [hunk 9]
```diff
@@ -1239,10 +1243,36 @@ let walletElementPaymentType: array<CardThemeType.mode> = [
...
+  let options =
+    list
+    ->getDictFromJson
+    ->getOptionalArrayFromDict(country)
+    ->Option.getOr([])
+
+  let val = options->Array.find(item =>
+    item
+    ->getDictFromJson
+    ->Dict.get("code")
```
Comment (PritishBudhiraja) [inline src/Utilities/Utils.res:1264]: getString

File: src/Utilities/Utils.res [hunk 10]
```diff
@@ -1239,10 +1243,36 @@ let walletElementPaymentType: array<CardThemeType.mode> = [
...
+    ->Option.flatMap(JSON.Decode.string)
+    ->Option.getOr("") === stateCode
+  )
+
+  switch val {
+  | Some(stateObj) =>
+    stateObj
+    ->getDictFromJson
+    ->Dict.get("name")
```
Comment (PritishBudhiraja) [inline src/Utilities/Utils.res:1273]: getString

File: src/Types/GooglePayType.res [hunk 11]
```diff
@@ -125,3 +141,18 @@ let jsonToPaymentRequestDataType: (paymentDataRequest, Dict.t<JSON.t>) => paymen
 
   paymentRequest
 }
+
+let billingContactItemToObjMapper = dict => {
+  {
```
Comment (PritishBudhiraja) [inline src/Types/GooglePayType.res:146]: already utils are open

File: src/Payments/PaypalSDK.res [hunk 12]
```diff
@@ -96,10 +113,25 @@ let make = (~sessionObj: SessionsType.token) => {
...
+
+                              let requiredFieldsBody = DynamicFieldsUtils.getPaypalRequiredFields(
+                                ~details=payload.details,
+                                ~paymentMethodTypes,
+                                ~statesList=stateJson,
+                              )
+
+                              let paypalBody =
+                                body
+                                ->Dict.fromArray
```
Comment (PritishBudhiraja) [inline src/Payments/PaypalSDK.res:125]: getJsonFromArrayOfJson

**Review by PritishBudhiraja — APPROVED** [review]
Code wise looks fine

**Review by prafulkoppalkar — APPROVED** [review]

**Pull request merged**